### ISSUE TITLE
Boss announcements

### ISSUE THREAD

**Issue by UndeadGamingCentral** [garbagemule/MobArena#504]
# Feature request
Separate boss announcements.
**Short description**
Being it so we have killing a boss read "Player has killed [boss name]" instead of just "Player has killed the boss!"
At the moment we can only have killing a boss or another announcement be one selection, which I am requesting we have it per arena in the arena config.

**Implementation details**
Per-arena announcements

**Comment by garbagemule** [comment]
I think this is much too narrow in scope. Why should boss announcements be any different than other announcements?

We can do better. 

Can you rephrase the feature request to something that encapsulates the announcement system as a whole instead of just one specific announcement? It seems as though the announcement system is due for a rework that allows for more than one variable per announcement, and perhaps arena-specific announcements. It would probably tie in well with a config-file split, where each arena has its own set of config-files for overriding a set of global/default config-files.

Let's start small but a tad more generalized.

**Comment by UndeadGamingCentral** [comment]
Adding this will help servers with multiple arenas setup seperate announcements per arena. Letting players know what boss was fight, died to, killed and more. As for a config-file split, that would help better find arenas to edit instead of looking through 1 config file to find an arena to edit.

**Comment by garbagemule** [comment]
Okay, here is an example of how a suggestion like this could be sufficiently detailed, such that someone might actually begin to work on an implementation.

The important thing here is the attention to user-facing detail. It isn't really all that interesting for server-owners to know about implementation details, but it _is_ necessary for server-owners to have opinions about how they would like to configure stuff. If you say "I'd like it if creepers could drop items", you aren't specifying how you would like to set things up to enable that, which means the implementation details _dictate_ the user-facing part of things. Developers will naturally work bottom-up if they aren't constrained or directed otherwise.

Another important thing is that this is not difficult. It isn't something "only a developer can do", and it isn't something that requires any kind of "skill". It just requires effort. Some people are "lazy", others are "busy". It doesn't matter. If you want a developer to invest effort into your suggestion, the least you can do is show that developer the same respect by investing effort into it yourself.

---

# Problem
The underlying problem here has nothing to do with bosses or boss announcements. It has to do with the rigidity of the announcements system and the limitations resulting hereof. What we need is flexibility to choose what details we would like to include in each announcement.

As an example, take the announcement `player-died`, which has the dead player's name as its only variable. It could be useful to also announce the cause of death, or perhaps for how long they survived before they died.

Another example is `wave-boss-ability`, which has the boss ability as its only variable. Because bosses can have custom names, it makes sense to have two variables in this announcement: the ability name _and_ the boss name.

# Solution
An immediately obvious solution is to introduce _named variables_, such that instead of the template being `Boss used ability: %!`, it could be `Boss used ability: %ability%!`. The variable symbols could be something else like angled brackets `<ability>` or curly braces `{ability}`, whichever works the best (**note:** server-owners deal with many plugins, so if consistency across plugins is a desired property, this is where you can truly pitch in and provide syntax examples from other plugins you like).

Directly implementing this would break all existing announcement configurations, so perhaps it would be necessary with an automatic conversion in MobArena itself, or perhaps an external conversion tool could be made (like when the announcements system was first reworked years ago). Alternatively, there could be a "primary variable" in each announcement for some time (maybe a couple of months), and using the old `%` symbol would produce a warning in the console, urging people to update their configurations.

# Call for feedback/discussion
- How do we know which variables are available in a given announcement?
- Should we just have a huge table of all announcements and their possible variables and descriptions of each variable?
- Or should we have a list of globally available variables and some specific variables per announcement?
- The variable `<player>` is probably understood as the player name. But what does `<wave>` denote? How do we decide on good, unambiguous variable names?
